RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc

```diff
@@ -0,0 +1,123 @@
+    pp.push_back(it2->clone());
+    // assign updated vector of GEMRecHits to reference id
+    ensembleRH[id.rawId()]=pp;
+  }
+
+  // Loop on the entire map <ref id, vector of GEMRecHits>
+  for(auto enIt=ensembleRH.begin(); enIt != ensembleRH.end(); ++enIt) {
+    
+    std::vector<const GEMRecHit*> gemRecHits;
+    std::map<uint32_t,const GEMEtaPartition* > ens;
+
+    std::vector<GEMRecHit* > pp = enIt->second;
+    std::vector<GEMRecHit*>::iterator ppit = pp.begin();
+    GEMRecHit * pphit = (*ppit);
+    // !!! important !!! for GE2/1 make that the chamber is always in station 3
+    int chambidstat = pphit->gemId().station(); if(chambidstat==2) chambidstat=3;
+    // would layer = 0 work? actually it should ...
+    GEMDetId chamberid = GEMDetId(pphit->gemId().region(), 1, chambidstat, 0, pphit->gemId().chamber(), 0); 
+    const GEMSuperChamber* chamber = geom_->superChamber(chamberid);
+    for(auto rechit = enIt->second.begin(); rechit != enIt->second.end(); ++rechit) {
+      gemRecHits.push_back(*rechit);
+      ens[(*rechit)->gemId()]=geom_->etaPartition((*rechit)->gemId());
+    }    
+
+    #ifdef EDM_ML_DEBUG // have lines below only compiled when in debug mode
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] -----------------------------------------------------------------------------"; 
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] found " << gemRecHits.size() << " rechits in GEM Super Chamber " << chamber->id()<<" ::"; 
+    for (auto rh=gemRecHits.begin(); rh!=gemRecHits.end(); ++rh){
+      auto gemid = (*rh)->gemId();
+      // auto rhr = gemGeom->etaPartition(gemid);
+      auto rhLP = (*rh)->localPosition();
+      // auto rhGP = rhr->toGlobal(rhLP);
+      // no sense to print local y because local y here is in the roll reference frame
+      // in the roll reference frame the local y of a rechit is always the middle of the roll, and hence equal to 0.0
+      edm::LogVerbatim("GEMSegmentBuilder") << "[RecHit :: Loc x = "<<std::showpos<<std::setw(9)<<rhLP.x() /*<<" Loc y = "<<std::showpos<<std::setw(9)<<rhLP.y()*/
+					    <<" BX = "<<(*rh)->BunchX()<<" -- "<<gemid.rawId()<<" = "<<gemid<<" ]";
+    }
+    #endif
+
+
+    GEMSegmentAlgorithm::GEMEnsemble ensemble(std::pair<const GEMSuperChamber*,      std::map<uint32_t,const GEMEtaPartition*> >(chamber,ens));
+    
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] run the segment reconstruction algorithm now";
+    // given the superchamber select the appropriate algo... and run it
+    std::vector<GEMSegment> segv = algo->run(ensemble, gemRecHits);
+    GEMDetId mid(enIt->first);
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] found " << segv.size() << " segments in GEM Super Chamber " << mid;
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] -----------------------------------------------------------------------------"; 
```

Comment:
It looks like this will be called very frequently per event.
This is too costly. Please change to LogTrace or hide behind a configurable ``` if (debug ) ``` or similar